### PR TITLE
rpc: retain information about failed connections

### PR DESCRIPTION
**Pull request by tbg**
Prior to this change, `rpc.Context` did not remember a connection after it
failed. A connection attempt would be created whenever we didn't have a
functioning connection to the given `(nodeID, targetAddress, class)` and
callers would be multiplexed onto it as long as that attempt was not known to
have failed.

The main problem with this setup was that it makes it difficult to provide good
observability because a remote node that just hasn't been dialed yet and one
that is down but between attempts looks the same. We could write code that
periodically tries to fully connect all nodes in the cluster to each other, but
even then exporting good metrics is challenging and while we're currently
comfortable with a densely connected cluster, that may change over time.

An additional challenge was presented by circuit breaking and logging of
attempts. Without state retained, we were limited to a simple retry scheme with
lots of redundant logging. It wasn't easy to log how long a connection had been
unhealthy (as an unhealthy connection was represented as an absent connection),
so in effect folks had to trawl through logs to grab the corresponding
timestamps of first and last failure.

Another piece of complexity were the RPC circuit breakers. These were
implemented at the NodeDialer-level (i.e. one circuit breaker per
`(NodeID,Class`)) but kept in the `rpc.Context` (which generally operates on
`(NodeID, Addr, Class)` because gossip also used them. The library they were
using uses recruitment of client traffic, which also isn't ideal as it could
periodically subject a SQL query to a timeout failure. We'd really prefer it
if the breakers probed without user traffic recruitment.

This PR addresses the above shortcomings:

- state is now retained across restarts on a `(NodeID, Addr, Class)` basis in
  the `rpc.Context. This has a few complications, for example we need to handle
decommissioned nodes, as well as nodes that restart under a new listening
address.
- the NodeDialer-level circuit breakers are removed.
- we're no longer recruiting user traffic for probes. Instead, we adopt the
  `util/circuit` package already used for Replica-level circuit breaking. This
library uses an async background probe to determine when to heal the breaker.

I also think the amount of incidental complexity in the `rpc` package is greatly
reduced as a result of this work.

I ran the `failover/non-system/` suite of tests to validate this change. The results
look good[^res].

[^res]: https://gist.github.com/tbg/5f5c716fa2fe5a21eca4a0c3a4a35069

Epic: CRDB-21710
Release note (ui change): a "Networking" metrics dashboard was added.
Release note (ops change): CockroachDB now exposes the following metrics:
- `rpc.connection.{un,}healthy`: Gauge of healthy/unhealthy connections,
- `rpc.connection.{un,}healthy_nanos`: Gauge of aggregate duration the currently {un,}healthy connections have been {,un}healthy for,
-  `grpc.connection.avg_round_trip_latency`: sum of weighted moving averages of round-trip latencies for all peers (it really only makes sense to consume these on a per-peer basis via prometheus)
- Counters `rpc.connection.heartbeats` and `rpc.connection.failures`.
When the `server.child_metrics.enabled` cluster setting is set, these metrics export per-peer values, meaning they can be used to derive a connectivity graph.
Release note (ops change): logging for intra-node RPC connection failures has improved by reducing frequently and improving information content.

### CODE REVIEW

**Comment by tbg** [comment]
Sneak Peek of metrics

image

Here are the prometheus metrics, with <source,dest> labels (needs `set cluster setting server.child_metrics.enabled = true;`)

https://gist.github.com/tbg/c7ec86c0d180580f17756a51fb5944e4

File: pkg/rpc/breaker_new_probe.go
Comment (aliher1911) [inline pkg/rpc/breaker_new_probe.go:156]: Why do we touch all peers here? How does failure of this connection link us to other potential peers?

File: pkg/rpc/context.go
Comment (aliher1911) [inline pkg/rpc/context.go:2776]: Needs to be named *Locked to show that we must lock peers prior to calling this as `shouldDeleteAfter` does.

File: pkg/rpc/context.go
Comment (aliher1911) [inline pkg/rpc/context.go:2820]: Is it ok that we could wait for sibling connection to get its health resolved here? Wouldn't that delay resolving gatedCC below.

**Review by aliher1911 — COMMENTED** [review]
Some initial questions on the first pass of new breakers.

File: pkg/rpc/context.go
Comment (tbg) [reply pkg/rpc/context.go:2401]: You can read a `nil` map just fine, the call you allude to bottoms out here:

```go
func (m *peerMap) getRLocked(k peerKey) (PeerSnap, bool) {
	p, ok := m.mu.m[k]
	if !ok {
		return PeerSnap{}, false
	}
	return p.snap(), true
}
```

That works fine, even though `m.mu.m` is nil. We basically make the map right as we're about to do the first write to it (which will panic on a nil map).

I can also "always" make the map in `NewContext`, maybe that's easier. The pattern I'm using here is more useful when the map remains unused most of the time, but here it's definitely going to be used. Your call.

File: pkg/rpc/breaker_new_probe.go
Comment (tbg) [reply pkg/rpc/breaker_new_probe.go:156]: It's just a convenient trigger, added this comment (with next push)

	// There might be other peers in the map that are pending deletion, but which
	// are no longer seeing activity. To eventually clear them out, we check all
	// conns when any conn fails. This avoids the need to have an extra goroutine
	// sitting in all of the goroutine stacks we look at during support.
	touchOldPeers(p.peers, now)

File: pkg/rpc/context.go
Comment (tbg) [reply pkg/rpc/context.go:2419]: Yeah just for ease of navigation. Let me know if you have a comment in mind that would help (one that isn't too long!) or whether you think this is not necessary.

File: pkg/rpc/context.go
Comment (aliher1911) [reply pkg/rpc/context.go:2820]: Maybe `rpc.Connection.wait` needs a better name then because that got me confused since I didn't look on the implementation. Something like `waitOrDefault`?

File: pkg/rpc/context.go
Comment (tbg) [reply pkg/rpc/context.go:2419]: Removed it.

File: pkg/rpc/context.go
Comment (tbg) [reply pkg/rpc/context.go]: Good idea, done.

File: pkg/rpc/context_test.go
Comment (aliher1911) [inline pkg/rpc/context_test.go:2610]: `(healthy > 0) != (hf != 0)` is not obvious at all. Not sure we can simplify the condition, but at least a comment saying that `checkDurations` mean HealthyFor, UnhealthyFor and RoundTrip must be non zero if corresponding counters are non zero.

**Review by aliher1911 — COMMENTED** [review]
I like how it is now looking overall, but the fact that we have so many "connection" entities is odd. It is not different from existing design though. What I think odd is that we have context that gives us connection on dial but then you need to connect to get actual connection. In socket API's you can get unconnected/unbound socket that you can connect, but you still use the same object. In our case we get potentially disconnected connection from context/dialer and then we connect and get grpc connection which is bound to this connection object.
It looks more like a factory of some sort as it doesn't provide any functionality that connection imply i.e. sending data.

Comments on `func (n *Dialer) dial(` seem outdated as its lost ins built in breaker functionality, but still mentions nil where argument was replaced by a bool. Maybe it should just describe things on higher level without going into breaker detail?

**Comment by aliher1911** [comment]
Mentioned briefly offline, but worth spelling here for posterity.

I think using cirtuit.Breaker to drive probe lifecycle is odd. It adds unnecessary complexity with callbacks. I think it would be easier to just start a probe when connection is established and let it inform breaker which will be used without its asyncStartProbe part. We can still get standardized logging events and state tracking out of it.

File: pkg/rpc/context_test.go
Comment (tbg) [reply pkg/rpc/context_test.go:2610]: I added this comment:

	// HealthyFor is nonzero if and only if there is at least one healthy connection, since it's
	// the sum of the connections' HealthyFor, and on a connection this is nonzero if and only
	// if the connection is healthy.

Does that help or do you think further clarification is necessary? The condition is exact, so all we can do is rewrite it. `hf > 0` is probably better than `hf != 0`, I made that change.

File: pkg/rpc/context.go
Comment (tbg) [reply pkg/rpc/context.go:314]: Yes, it's just a future. I renamed it to `connFuture`, that's a better name, thanks for the hint.

File: pkg/rpc/testdata/TestReconnection/initial.txt
Comment (tbg) [reply pkg/rpc/testdata/TestReconnection/initial.txt:33]: I added some comments, not sure this is what you had in mind, PTAL.

File: pkg/rpc/datadriven_test.go
Comment (tbg) [reply pkg/rpc/datadriven_test.go:61]: I am about to push a significant update to the datadriven test, see main thread.

**Comment by tbg** [comment]
Pushed another update - improving the datadriven test and fixing bugs/adding logging that came out of that. I feel pretty good about all of this working now.

The first new commit is:

> 33e89b29f7e TODO(tbg): review the preceding commits, I think they all amend into the wrong one.

File: pkg/rpc/context_test.go
Comment (aliher1911) [inline pkg/rpc/context_test.go:1042]: This test is still referenced in comments in this file grpcRunKeepaliveTestCase()

File: pkg/rpc/datadriven_test.go
Comment (aliher1911) [inline pkg/rpc/datadriven_test.go:80]:
```
d.MaybeScanArgs(t, "decommissioned", &dc)
```

File: pkg/rpc/datadriven_test.go
Comment (aliher1911) [inline pkg/rpc/datadriven_test.go:97]: MaybeScanArgs as well

File: pkg/rpc/datadriven_test.go
Comment (aliher1911) [reply pkg/rpc/datadriven_test.go:55]: Is it out of date already? 🤣  Maybe mention n1'' convention? Looks neat in test data but magical until you see that we have preset 2 servers with same name that you skip.

**Review by aliher1911 — COMMENTED** [review]
Some comments about tests otherwise looks great!

File: pkg/rpc/datadriven_test.go
Comment (tbg) [reply pkg/rpc/datadriven_test.go:55]: Fixed in next push :)

File: pkg/rpc/peer_test.go
Comment (aliher1911) [inline pkg/rpc/peer_test.go:150]: Fluent map builder is fancy, but isn't `pmb(defSelf(), p(n,t,c,h))` as a func with varargs more readable because less () and {}?
I like simple, but YMMV so up to you.

**Review by aliher1911 — APPROVED** [review]
🎉🎉🎉 LGTM 🎉🎉🎉 
🍿

File: pkg/rpc/peer_test.go
Comment (tbg) [reply pkg/rpc/peer_test.go:150]: You're right, but it's a little awkward because the mock `peer` I construct here doesn't know its own `connKey`. I still think your suggestion is directionally right, but I'm going to leave as is.

**Comment by tbg** [comment]
bors r=aliher1911

**Comment by tbg** [comment]
Thanks I'll take a look tomorrow!

**Comment by tbg** [comment]
This should hopefully fix it once merged: https://github.com/cockroachdb/cockroach/pull/104945

**Comment by data-matt** [comment]
Folks, any chance of backporting this flag to 22.2? thanks

**Comment by tbg** [comment]
No, this is a major change. We won't be able to backport it.

**Comment by data-matt** [comment]
@tbg , the cluster setting `server.child_metrics.enabled` will be in 23.1+ ? I see references to 23.1 here.

Thanks cc @florence-crl

**Comment by erikgrinaker** [comment]
>@tbg , the cluster setting server.child_metrics.enabled will be in 23.1+ ? I see references to 23.1 here.

The setting `server.child_metrics.enabled` has existed for years, and will enable additional Prometheus metrics for possibly-high-cardinality metrics. For example, changefeeds appear to use them to expose additional per-changefeed metrics.

However, prior to 23.2, this setting does not expose additional RPC metrics. In 23.2, it will also expose per-peer connection metrics as described in the release notes in this PR. This does not happen in 23.1, and can not since the necessary changes are too invasive.

The 23.1 reference you see here was for a minor bugfix that was extracted out of this PR, see https://github.com/cockroachdb/cockroach/pull/105626. It is not related to the functionality you're interested in.

**Comment by data-matt** [comment]
Thanks for the clear answer @erikgrinaker.

**Pull request merged**
